### PR TITLE
`/pep` autocomplete

### PR DESCRIPTION
**Pull request by Endercheif**

### CODE REVIEW

**Comment by Endercheif** [comment]
`poetry.lock` :sob: how to fix

**Comment by Endercheif** [comment]
Implements #55

File: mcodingbot/plugins/peps.py [hunk 1]
```diff
@@ -30,6 +30,23 @@ def get_dismiss_button(id: hikari.Snowflake) -> hikari.api.ActionRowBuilder:
     return action_row
 
 
+async def autocomplete_pep(
+    ctx: crescent.AutocompleteContext,
+    options: hikari.AutocompleteInteractionOption,
```
Comment (Lunarmagpie) [inline mcodingbot/plugins/peps.py:45]: should be `option` (no s)

File: mcodingbot/plugins/peps.py [hunk 2]
```diff
@@ -41,17 +58,14 @@ async def update_peps() -> None:
     name="pep", description="Find a Python Enhancement Proposal."
 )
 class PEPCommand:
-    pep_number = crescent.option(
-        int, "The PEP number.", name="pep-number", min_value=0, max_value=9999
-    )
-    show_embed = crescent.option(
-        bool, "Whether to show the embed.", name="show-embed", default=True
+    pep = crescent.option(
+        int, "PEP number/title", autocomplete=autocomplete_pep
```
Comment (Lunarmagpie) [inline mcodingbot/plugins/peps.py:72]:
```suggestion
        int, "The PEP number or title.", autocomplete=autocomplete_pep
```

File: mcodingbot/utils/peps.py [hunk 3]
```diff
@@ -2,12 +2,13 @@
...
 from dataclasses import dataclass
 from logging import getLogger
-from typing import TYPE_CHECKING, Sequence
+from typing import TYPE_CHECKING, Sequence, Generator
 
 import aiohttp
 import hikari
 
 from mcodingbot.config import CONFIG
+from mcodingbot.utils.search import find
```
Comment (Lunarmagpie) [inline mcodingbot/utils/peps.py:11]:
```suggestion
from mcodingbot.utils import find
```
Its re-exported

File: mcodingbot/utils/peps.py [hunk 4]
```diff
@@ -39,10 +41,19 @@ async def fetch_pep_info(self, bot: Bot) -> None:
...
+                self._pep_map = {
+                    k: f"{v.title} ({v.number})" for k, v in self._peps.items()
+                }
 
     def get(self, pep_number: int) -> PEPInfo | None:
         return self._peps.get(pep_number)
 
+    def search(
+        self, query: str, limit: int = 20
+    ) -> Generator[PEPInfo, None, None]:
```
Comment (Lunarmagpie) [inline mcodingbot/utils/peps.py:53]: This returns `typing.Iterable[PEPInfo]`

File: mcodingbot/utils/peps.py [hunk 5]
```diff
@@ -39,10 +41,19 @@ async def fetch_pep_info(self, bot: Bot) -> None:
                     )
                     for pep_id, pep in (await resp.json()).items()
                 }
+                self._pep_map = {
+                    k: f"{v.title} ({v.number})" for k, v in self._peps.items()
+                }
```
Comment (Lunarmagpie) [inline mcodingbot/utils/peps.py:46]: Gonna need to think about this. I feel like this can be done without a second dict.

**Review by Lunarmagpie — CHANGES_REQUESTED** [review]

File: mcodingbot/utils/peps.py (hunk 4 above)
Comment (CircuitSacul) [reply mcodingbot/utils/peps.py:53]: No it returns a generator. Generators are also Iterables, but it's best to be specific with returns types and broad with input types.

File: mcodingbot/utils/peps.py (hunk 5 above)
Comment (CircuitSacul) [reply mcodingbot/utils/peps.py:46]: The search function accepts a very specific format (either Collection[str] or Mapping[value, str]). Passing the original dict wouldn't work, and passing just the title would make it impossible to get the rest of the pep info from the response.

The dict either needs to be generated every time this function is called, or stored separately.

I could be missing something ofc.

File: mcodingbot/utils/peps.py (hunk 4 above)
Comment (Endercheif) [reply mcodingbot/utils/peps.py:53]: Keep it or change it?

File: mcodingbot/utils/search.py [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+from collections.abc import Mapping
+from typing import TypeVar
+
+from rapidfuzz import fuzz, process
+
+K = TypeVar("K")
+V = TypeVar("V")
+
+
+def find(
```
Comment (CircuitSacul) [inline mcodingbot/utils/search.py:12]: I feel like we need a more descriptive name maybe. Perhaps `fuzzysearch`?

File: mcodingbot/utils/peps.py (hunk 4 above)
Comment (CircuitSacul) [reply mcodingbot/utils/peps.py:53]: I say keep it, unless @Lunarmagpie has a good reason to change it

File: mcodingbot/utils/peps.py (hunk 5 above)
Comment (Endercheif) [reply mcodingbot/utils/peps.py:46]: I spent a while on thinking of a better way to do this. Unfortunately this is the best way other than generating it every time it is called.

File: mcodingbot/utils/peps.py [hunk 7]
```diff
@@ -39,10 +41,19 @@ async def fetch_pep_info(self, bot: Bot) -> None:
...
 
     def get(self, pep_number: int) -> PEPInfo | None:
         return self._peps.get(pep_number)
 
+    def search(
+        self, query: str, limit: int = 20
+    ) -> Generator[PEPInfo, None, None]:
+        res = find(query, self._pep_map, limit=limit)
+        return (self.get(pep[2]) for pep in res)
```
Comment (CircuitSacul) [inline mcodingbot/utils/peps.py:55]: Don't forget to fix this part (this is why mypy is failing). One way to fix it is like this:

```py
return (
    pep_info
    for pep in res
    if (pep_info := self.get(pep[2])) is not None
)

File: mcodingbot/utils/peps.py (hunk 4 above)
Comment (Lunarmagpie) [reply mcodingbot/utils/peps.py:53]: Generator inherits from Iterable (check typeshed)
we dont need any of the added things. i do not believe they apply when its Generator[T, None, None]

Comment (CircuitSacul) [reply mcodingbot/utils/peps.py]: If generator inherits from Iterable then this should be a Generator, as that's what's being returned.

You wouldn't do this
```py
def function() -> object:
    return "hello"
```

You would use `str`, which is a subclass of `object`, as that is more specific. Your return types should be as specific as possible, and argument types should be as broad as possible.

Comment (Lunarmagpie) [reply mcodingbot/utils/peps.py:53]: nobody returns a string to use it like an object...
i will correct myself `Iterator` is more accurate. `Generator[T, None, None`] does not give us a single thing that does shit over Iterator.  I'm sure throw, close, gi_code, gi_frame, gi_yieldfrom, and gi_running are real useful. Your argument assumes they are as useful as basic string methods.

Comment (CircuitSacul) [reply mcodingbot/utils/peps.py:53]: Generators behave completely differently from all other Iterables, as Generators are consumable.

Comment (CircuitSacul) [reply mcodingbot/utils/peps.py:53]: https://discord.com/channels/826875707418214451/851854179534372884/1019844894912675892

Comment (Endercheif) [reply mcodingbot/utils/peps.py:53]: @trag1c thoughts ~~save me~~

Comment (Lunarmagpie) [reply mcodingbot/utils/peps.py:53]: > 

So use iterator. The docs recommend it as an alternative as well.

File: mcodingbot/utils/search.py [hunk 8]
```diff
@@ -0,0 +1,24 @@
+from __future__ import annotations
+
+from collections.abc import Mapping
```
Comment (Lunarmagpie) [inline mcodingbot/utils/search.py:3]: Use typing mapping for consistentcy

**Review by Lunarmagpie — CHANGES_REQUESTED** [review]

File: mcodingbot/utils/peps.py (hunk 4 above)
Comment (trag1c) [reply mcodingbot/utils/peps.py:53]: > @trag1c thoughts? ~save me~

Use `Iterator[PEPInfo]`

File: mcodingbot/utils/peps.py [hunk 9]
```diff
@@ -39,10 +41,23 @@ async def fetch_pep_info(self, bot: Bot) -> None:
...
 
+    def search(
+        self, query: str, limit: int = 20
+    ) -> Generator[PEPInfo, None, None]:
+        res = fuzzy_search(query, self._pep_map, limit=limit)
+        return (
+            pep_info
+            for pep in res
+            if (pep_info := self.get(pep[2])) is not None
+        )
```
Comment (trag1c) [inline mcodingbot/utils/peps.py:59]: Why not just use `yield` in a normal loop? Would be more readable imo (mainly because of the walrus)

Comment (Lunarmagpie) [reply mcodingbot/utils/peps.py:59]: +1

**Review by Lunarmagpie — CHANGES_REQUESTED** [review]

File: mcodingbot/utils/peps.py (hunk 9 above)
Comment (Lunarmagpie) [reply mcodingbot/utils/peps.py:59]: +1 for tragic

Comment (CircuitSacul) [reply mcodingbot/utils/peps.py:59]: A yield seems like a good idea.

```py
def search(...) -> Generator[PEPInfo, None, None]:
    res = ...
    for pep in res:
        if (pep_info := self.get(pep[2])):
            yield pep
```

**Review by Lunarmagpie — APPROVED** [review]

**Review by CircuitSacul — APPROVED** [review]

File: mcodingbot/utils/search.py [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+
+from typing import Mapping, TypeVar
+
+from rapidfuzz import fuzz, process
+
+K = TypeVar("K")
+V = TypeVar("V")
+
+
+def fuzzy_search(
```
Comment (Lunarmagpie) [inline mcodingbot/utils/search.py:11]: Removes default values being assigned in different places. They are already different so it's an issue.
```python
def fuzzy_search(
    query: str,
    choices: Mapping[K, V],
    *,
    score_cutoff: int | float | None,
    limit: int,
) -> list[tuple[V, float, K]]:
```

Comment (Endercheif) [reply mcodingbot/utils/search.py:11]: can you make this into a suggestion pls?

Comment (Lunarmagpie) [reply mcodingbot/utils/search.py:11]: No. I have locked myself out of the room with my computer so I'm a bit fucked rn.

File: mcodingbot/utils/search.py [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+    query: str,
+    choices: Mapping[K, V],
+    score_cutoff: int | float | None = None,
+    limit: int = 10,
+) -> list[tuple[V, float, K]]:
```
Comment (trag1c) [inline mcodingbot/utils/search.py:16]:
```suggestion
def fuzzy_search(
    query: str,
    choices: Mapping[K, V],
    *,
    score_cutoff: int | float | None,
    limit: int,
) -> list[tuple[V, float, K]]:
```

File: mcodingbot/utils/peps.py [hunk 12]
```diff
@@ -39,10 +41,19 @@ async def fetch_pep_info(self, bot: Bot) -> None:
...
                     for pep_id, pep in (await resp.json()).items()
                 }
+                self._pep_map = {
+                    k: f"{v.title} ({v.number})" for k, v in self._peps.items()
+                }
 
     def get(self, pep_number: int) -> PEPInfo | None:
         return self._peps.get(pep_number)
 
+    def search(self, query: str, limit: int = 20) -> Iterator[PEPInfo]:
```
Comment (Lunarmagpie) [inline mcodingbot/utils/peps.py:51]:
```suggestion
    def search(self, query: str, *, limit: int) -> Iterator[PEPInfo]:
```

File: mcodingbot/plugins/peps.py [hunk 13]
```diff
@@ -40,6 +40,23 @@ def get_dismiss_button(id: hikari.Snowflake) -> hikari.api.ActionRowBuilder:
...
 
 
+async def autocomplete_pep(
+    ctx: crescent.AutocompleteContext,
+    option: hikari.AutocompleteInteractionOption,
+) -> list[hikari.CommandChoice]:
+    if not option.value:
+        return []
+
+    results = pep_manager.search(option.value)
```
Comment (Lunarmagpie) [inline mcodingbot/plugins/peps.py:50]:
```suggestion
    results = pep_manager.search(option.value, limit=20)
```

**Review by Lunarmagpie — CHANGES_REQUESTED** [review]
I don't think it makes more sense to abstract the limit out of the autocomplete function.

File: mcodingbot/utils/search.py [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+K = TypeVar("K")
+V = TypeVar("V")
+
+
+def fuzzy_search(
+    query: str,
+    choices: Mapping[K, V],
+    *,
+    score_cutoff: int | float | None,
+    limit: int,
```
Comment (CircuitSacul) [inline mcodingbot/utils/search.py:16]: kw-only arguments should have a default value imo.

Just make both of these None, I think?

Comment (CircuitSacul) [reply mcodingbot/utils/search.py:16]: wait lunar just asked you to change that lol. hm

Comment (CircuitSacul) [reply mcodingbot/utils/search.py:16]: okay yeah if you can make this arguments None that would be great.

Comment (Endercheif) [reply mcodingbot/utils/search.py:16]: @CircuitSacul leave a suggestion then i will change it

**Review by Lunarmagpie — APPROVED** [review]

**Review by trag1c — APPROVED** [review]

**Comment by Endercheif** [comment]
🚀🚀🚀

**Pull request merged**